### PR TITLE
Feature/issue 467 go generate on build

### PR DESCRIPTION
**Pull request by jgimeno**
Now the build checks that the go generated files are up to date, if not it fails.

Closes #

https://github.com/kowala-tech/kcoin/issues/467

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
...
+	go generate ./client/internal/jsre/deps/
+	go generate ./client/knode/
+	go generate ./client/knode/tracers/internal/tracers/
+	go generate ./client/p2p/discv5/
+	go generate ./notifications/blockchain/
+	go generate ./notifications/environment/
+	go generate ./notifications/keyvalue/
+	go generate ./notifications/notifier/
+	go generate ./notifications/protocolbuffer/
+	go generate ./wallet-backend/protocolbuffer/
```
Comment (acroca) [inline Makefile]: I think we'll easily forget to keep this list up to date. What if we do like in the `bindings` task?

Something like:
```
egrep -ir --exclude-dir=vendor --exclude-dir=_workspace --include \*.go "//go:generate" . | grep -v client/contracts | sed -E 's/^([^\:]*).*$/\1/' | uniq
```

File: Makefile [hunk 2]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
 client/build/bin/abigen:
 	cd client; build/env.sh go run build/ci.go install ./cmd/abigen
 
+go-generate: moq go-bindata stringer gencodec mockery ensure-notifications ensure-wallet-backend
+	go get -u github.com/golang/protobuf/protoc-gen-go
```
Comment (acroca) [inline Makefile]: Move this to its own task at the end (with the other tools) and only install it if missing

File: .drone.yml [hunk 3]
```diff
@@ -10,13 +10,19 @@ clone:
 
 pipeline:
   bindings:
-    image: kowalatech/go:1.0.4
+    image: kowalatech/go:1.0.8
     commands:
       - make bindings
 
+  go-generate:
```
Comment (acroca) [inline .drone.yml]: I'd make this run on `push` only

File: Makefile [hunk 4]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
...
+	go generate ./client/p2p/discv5/
+	go generate ./notifications/blockchain/
+	go generate ./notifications/environment/
+	go generate ./notifications/keyvalue/
+	go generate ./notifications/notifier/
+	go generate ./notifications/protocolbuffer/
+	go generate ./wallet-backend/protocolbuffer/
+
+assert-no-generate:
+	git status
```
Comment (acroca) [inline Makefile]: Is this just to give some context to the error or a mistake?

File: Makefile [hunk 5]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
...
+	if ! git diff-index --quiet HEAD; then echo "There are uncommited go generate files."; exit 1; fi
+
+ensure-notifications: dep
+	cd notifications && \
+	$(GOPATH)/bin/dep ensure --vendor-only && \
+	cd ..
```
Comment (acroca) [inline Makefile]: I think you don't need to go back

File: Makefile (hunk 4 above)
Comment (jgimeno) [reply Makefile]: I prefer leaving this status just for getting context, when it fails it is not always easy to know what happened. So it is not harmful to have this.

File: Makefile [hunk 6]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
 client/build/bin/abigen:
 	cd client; build/env.sh go run build/ci.go install ./cmd/abigen
 
+go-generate: moq go-bindata stringer gencodec mockery ensure-notifications ensure-wallet-backend
```
Comment (acroca) [inline Makefile]: The makefile tasks have to use underscores, so `go_generate`

File: Makefile [hunk 7]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
...
+	go generate ./client/knode/tracers/internal/tracers/
+	go generate ./client/p2p/discv5/
+	go generate ./notifications/blockchain/
+	go generate ./notifications/environment/
+	go generate ./notifications/keyvalue/
+	go generate ./notifications/notifier/
+	go generate ./notifications/protocolbuffer/
+	go generate ./wallet-backend/protocolbuffer/
+
+assert-no-generate:
```
Comment (acroca) [inline Makefile]: underscores

File: Makefile [hunk 8]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
...
+	git status
+	if ! git diff-index --quiet HEAD; then echo "There are uncommited go generate files."; exit 1; fi
+
+ensure-notifications: dep
```
Comment (acroca) [inline Makefile]: underscores

File: Makefile [hunk 9]
```diff
@@ -92,6 +92,38 @@ client/contracts/truffle/node_modules:
...
+	cd notifications && \
+	$(GOPATH)/bin/dep ensure --vendor-only && \
+	cd ..
+
+ensure-wallet-backend: dep
```
Comment (acroca) [inline Makefile]: underscores

File: Makefile (hunk 1 above)
Comment (acroca) [reply Makefile]: actually, if we combine this with the bindings, we can just `go generate ./...`

Comment (jgimeno) [reply Makefile]: go generate fails if bindings dont happen before the other, at least this was what I was getting on the beginning.

Comment (jgimeno) [reply Makefile]: I have been trying the approach you gave me, but the egrep version of the ci does not have that options. If you can help me in this case? I am not as good in bash!

@acroca

**Pull request merged**